### PR TITLE
ASR: split `Compare` based on different types

### PR DESCRIPTION
**Pull request by namannimmo10**

### CODE REVIEW

**Comment by namannimmo10** [comment]
This is almost ready. I'll raise a merge request in LFortran just to make sure that every test passes there. Then we'll merge both at once. For now, a review would be helpful!

**Comment by namannimmo10** [comment]
MR opened: https://gitlab.com/lfortran/lfortran/-/merge_requests/1784

**Comment by namannimmo10** [comment]
MR opened: https://gitlab.com/lfortran/lfortran/-/merge_requests/1784

File: src/libasr/ASR.asdl [hunk 1]
```diff
@@ -217,7 +217,6 @@ expr
         -- Such as: (x, y+z), (3.0, 2.0) generally not known at compile time
     | ComplexConstructor(expr re, expr im, ttype type, expr? value)
     | NamedExpr(expr target, expr value, ttype type)
-    | Compare(expr left, cmpop op, expr right, ttype type, expr? value, expr? overloaded)
```
Comment (certik) [inline src/libasr/ASR.asdl:220]: Also you need to put in `OverloadedCompare` node to represent the `expr? overloaded`.

**Comment by certik** [comment]
I think this change is in the right direction, I think we should do it. The comparison operator becomes simple/more explicit.

**Comment by namannimmo10** [comment]
@certik -
```python
from ltypes import c32, c64

def main():
  c: c32
  c = complex(3, 2)
  print(c)

  c2: c64
  c2 = complex(4.5, 6.7)
  print(c2)


main()
```
This code seems to produce the following ASR on the `main` branch:
```
(TranslationUnit 
   (SymbolTable 
      1 
      {
         _lpython_main_program: 
            (Subroutine 
               (SymbolTable 
                  100 
                  {
                  }) 
               _lpython_main_program [] [
               (SubroutineCall 1 main () [] ())] 
               Source 
               Public 
               Implementation () .false. .false.), 
         complex@__lpython_overloaded_5__complex: 
            (ExternalSymbol 
               1 
               complex@__lpython_overloaded_5__complex 4 __lpython_overloaded_5__complex 
               lpython_builtin [] 
               __lpython_overloaded_5__complex 
               Public), 
      complex@__lpython_overloaded_9__complex: 
         (ExternalSymbol 
            1 
            complex@__lpython_overloaded_9__complex 4 __lpython_overloaded_9__complex 
            lpython_builtin [] 
            __lpython_overloaded_9__complex 
            Public), 
   lpython_builtin: (IntrinsicModule lpython_builtin), 
main: 
   (Subroutine 
      (SymbolTable 
         2 
         {
            c: 
               (Variable 
                  2 
                  c 
                  Local () () 
                  Default 
                  (Complex 4 []) 
                  Source 
                  Public 
                  Required .false.), 
            c2: 
               (Variable 
                  2 
                  c2 
                  Local () () 
                  Default 
                  (Complex 8 []) 
                  Source 
                  Public 
                  Required .false.), 
            complex: 
               (ExternalSymbol 
                  2 
                  complex 4 complex 
                  lpython_builtin [] 
                  complex 
                  Private)
         }) 
      main [] [
      (= 
         (Var 2 c) 
         (Cast 
            (FunctionCall 1 complex@__lpython_overloaded_9__complex 2 complex [
               ((IntegerConstant 3 
                  (Integer 4 []))) 
               ((IntegerConstant 2 
                  (Integer 4 [])))] 
               (Complex 8 []) 
               (ComplexConstant   3.00000000000000000e+00   2.00000000000000000e+00 
                  (Complex 8 [])) ()) 
            ComplexToComplex 
            (Complex 4 []) 
            (ComplexConstant   3.00000000000000000e+00   2.00000000000000000e+00 
               (Complex 4 []))) ()) 
      (Print () [
         (Var 2 c)]) 
      (= 
         (Var 2 c2) 
         (FunctionCall 1 complex@__lpython_overloaded_5__complex 2 complex [
            (
               (RealConstant   4.50000000000000000e+00 
                  (Real 8 []))) 
            (
               (RealConstant   6.70000000000000018e+00 
                  (Real 8 [])))] 
            (Complex 8 []) 
            (ComplexConstant   4.50000000000000000e+00   6.70000000000000018e+00 
               (Complex 8 [])) ()) ()) 
      (Print () [
         (Var 2 c2)])] 
      Source 
      Public 
      Implementation () .false. .false.), 
main_program: 
   (Program 
      (SymbolTable 
         99 
         {
         }) 
      main_program [] [
      (SubroutineCall 1 _lpython_main_program () [] ())])
      }) 
   [])

```
It doesn't look correct to me -- there's some issue with casting.  And because of this, the tests don't pass.

**Comment by certik** [comment]
LPython should translate `x = complex(a, b)` into `Assign(ComplexConstructor(a, b))` and `complex(1.0, 3.0)` into `ComplexConstant(1.0, 3.0)`, it should not call a function `_lpython_overloaded_9__complex`, as it currently does.

**Comment by certik** [comment]
But that issue is unrelated --- we can make that change later.

For this PR we should split Compare, but otherwise not do any other modifications.

**Comment by Smit-create** [comment]
> It doesn't look correct to me -- there's some issue with casting. And because of this, the tests don't pass.

What's the issue here? I see that it works correctly. Any pointers?

> LPython should translate `x = complex(a, b)` into `Assign(ComplexConstructor(a, b))` and `complex(1.0, 3.0)` into `ComplexConstant(1.0, 3.0)`, it should not call a function `_lpython_overloaded_9__complex`, as it currently does.

Yeah, it uses `complex` from `lpython_builtin` but doesn't create any bug I suppose?

File: src/libasr/ASR.asdl [hunk 2]
```diff
@@ -266,6 +270,7 @@ expr
     | ArrayPack(expr array, expr mask, expr? vector, ttype type, expr? value)
     | Transfer(expr source, expr mold, expr? size, ttype type, expr? value)
     | DerivedRef(expr v, symbol m, ttype type, expr? value)
+    | OverloadedCompare(expr left, cmpop op, expr right, ttype type, expr? value, expr overloaded)
```
Comment (czgdp1807) [inline src/libasr/ASR.asdl:273]: I am -1 on putting overloaded compare as a separate node. An extra attribute in `Compare` was enough I think.

**Comment by czgdp1807** [comment]
I feel this is a very big change. May be first we should add new nodes and if it works correctly then remove the older nodes. Doing everything all at once makes reviewing harder and the predicting the long term effect of these changes is tricky. @namannimmo10 We should discuss a step by step strategy to achieve this kind of ASR refactoring.

File: src/libasr/ASR.asdl (hunk 2 above)
Comment (certik) [reply src/libasr/ASR.asdl:273]: We don't have Compare anymore.

So which of the IntegerCompare, RealCompare etc should have the extra member? Note that neither Python nor Fortran allows you to overload Integer or Real comparison. It allows you to overload a user defined type compare, which is what the OverloadedCompare node does.

Comment (czgdp1807) [reply src/libasr/ASR.asdl:273]: No issues. Let’s have it then.

**Comment by czgdp1807** [comment]
@namannimmo10 Could you please resolve the conflicts here anyways?

**Comment by certik** [comment]
It's a huge change, although perhaps smaller than the BinOp split change.

One could do it step by step, although one issue is that often errors only show after we actually remove the BinOp node, otherwise you don't easily notice which places in the code need updating.

The best workflow with these big changes is to get them done relatively quickly, with as little changes as possible, and make tests pass, and just get it done and get it in. Then we'll polish it as we go.

Alternatively, we can just keep BinOp and only split IntegerBinOp, but as I said, it might actually make it more difficult, as we will not see all the changes that need fixing.

**Comment by namannimmo10** [comment]
The CI is green. I suggest that we merge this PR as it is. Sorry, this was a big change! But this kind of refactoring had to be done by removing/adding the ASR nodes altogether.

**Comment by certik** [comment]
I agree to just merge. I am reviewing it now to see if there are any blockers.

**Review by certik — APPROVED** [review]
I think that this is fine. No major issues. Let's merge this.

**Comment by certik** [comment]
@namannimmo10 would you mind please sending a similar PR to LFortran?

**Comment by certik** [comment]
Thanks a lot for doing this. Big change, a lot of work to get everything done. I think the ASR looks a lot cleaner, so does the LLVM backend (our main backend). When LLVM looks clean, all other backends are typically also clean, by proxy.

**Comment by namannimmo10** [comment]
> @namannimmo10 would you mind please sending a similar PR to LFortran?

[!1784](https://gitlab.com/lfortran/lfortran/-/merge_requests/1784) was merged. ✌️

**Comment by certik** [comment]
Excellent! Thank you again for doing that. It's a lot of work.

In terms of ASR design, what else should we do, if anything, right now?

**Comment by namannimmo10** [comment]
My pleasure!

Some random ideas:

We should verify the ASR optimization passes since the significant splitting of nodes is done ([!709](https://gitlab.com/lfortran/lfortran/-/issues/709)).

Since we are on our way to having a self-sufficient ASR layer, we should try to avoid overloading functions, and have ASR nodes do the job. For example, see https://github.com/lcompilers/lpython/issues/410. As a result, the ASR looks a lot cleaner -- it is one of LCompilers' major selling points.

I was reading about the [Intermediate representations in `rustc`](https://rustc-dev-guide.rust-lang.org/overview.html#intermediate-representations). It looks like they have three IRs between AST to LLVM IR, each having a different purpose. We have only one such IR and we even do some high-level optimizations in it. I'm sure this is a great design as noticed from the benchmarks. But I'm curious, why didn't you decide to have more than one intermediate representation?

**Comment by czgdp1807** [comment]
> We should verify the ASR optimization passes since the significant splitting of nodes is done ([!709](https://gitlab.com/lfortran/lfortran/-/issues/709)).

I will start doing this.

**Comment by czgdp1807** [comment]
I think in `rustc` they have HIR, Typed HIR and Mid level HIR. ASR fulfills the purposes of each of these simultaneously i.e, type checking is easily possible in ASR and it is sort of a forest of trees (similar to control flow graph). Each stmt follows the previous and each stmt has expression trees. So I don’t feel the need to separate out ASR more or in other words I feel it is self sufficient concept. @certik Do you have plans to split out ASR into more intermediate representations?

**Comment by certik** [comment]
Yes, things like bool should be done in ASR itself I think.

With each new IR we add, there is a very likely compilation slow down (you can only avoid it if somehow it simplifies the backend so much that it actually becomes faster), as well as more code to maintain, etc. So adding a new IR must pay for itself. Adding ASR seems to pay for itself, some compilation slowdown might be there, it's hard to measure, but it greatly simplified the whole design.

We can sit down and brainstorm what other alternative designs there are and if any of them would be better. For example we can see how HIR, Typed HIR and Mid level HIR corresponds to ASR. For example, we will have a lot of optimization passes. What is the best IR to do optimization passes fast?

It seems the current ASR is fulfilling these roles very well, but I am always interested in seeing if there is a way to do it even better.

**Pull request merged**